aea/connections/base.py

```diff
@@ -301,3 +301,143 @@ def is_connecting(self) -> bool:  # pragma: nocover
+        """Remove task on done, log error if happened."""
+        self._tasks.remove(task)
+        if task.exception():
+            # cause task.get_name for python3.8+
+            task_name = getattr(task, "task_name", "TASK NAME NOT SET")
+            self.logger.exception(
+                f"in task `{task_name}`, exception occured", exc_info=task.exception(),
+            )
+
+    async def _run_in_pool(self, fn: Callable, *args) -> Any:
+        """Run sync function in threaded executor pool."""
+        return await self._loop.run_in_executor(self._executor_pool, fn, *args)
+
+    def put_envelope(self, envelope: Optional["Envelope"]) -> None:
+        """Put envelope in to the incoming queue."""
+        self._ensure_connected()
+        self._loop.call_soon_threadsafe(
+            self._incoming_messages_queue.put_nowait, envelope
+        )
+
+    async def connect(self) -> None:
+        """Connect connection."""
+        with self._connect_context():
+            self._loop = asyncio.get_event_loop()
+            self._incoming_messages_queue = asyncio.Queue()
+            self._set_executor_pool()
+            await self._run_in_pool(self.on_connect)
+        self.start_main()
+
+    async def disconnect(self) -> None:
+        """Disconnect connection."""
+        self._ensure_connected()
+        with self._state.transit(
+            initial=ConnectionStates.disconnecting,
+            success=ConnectionStates.disconnected,
+            fail=ConnectionStates.disconnected,
+        ):
+            if self._tasks:
+                await asyncio.wait(self._tasks)
+            await self._run_in_pool(self.on_disconnect)
+
+    async def send(self, envelope: "Envelope") -> None:
+        """Send envelope to connection."""
+        self._ensure_connected()
+        task = self._loop.create_task(self._run_in_pool(self._send_wrapper, envelope))
+        # cause task.set_name for python3.8+
+        setattr(task, "task_name", f"On send({envelope})")  # noqa
+        task.add_done_callback(self._task_done_callback)
+        self._tasks.add(task)
+
+    def _send_wrapper(self, *args) -> None:
+        """Check is connected and call on_send method."""
+        self._ensure_connected()
+        self.on_send(*args)
+
+    async def receive(self, *args, **kwargs) -> Optional["Envelope"]:
+        """Get an envelope from the connection."""
+        self._ensure_connected()
+        return await self._incoming_messages_queue.get()
+
+    def start_main(self) -> None:
+        """Start main function of the connection."""
+
+        def _():
+            task = self._loop.create_task(self._run_in_pool(self.main))
+            # cause task.set_name sfor python3.8+
+            setattr(task, "task_name", "Connection main")  # noqa
+            task.add_done_callback(self._task_done_callback)
+            self._tasks.add(task)
+
+        self._loop.call_soon_threadsafe(_)
+
+    def main(self) -> None:
```

Comment:
Add `abstractmethod` decorator?

Reply:
no, sometimes main is not required, see no reason to force develope to define empty main if not needed